aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -599,6 +599,98 @@ Tensor& diff_out(const Tensor& self, int64_t n, int64_t dim, const c10::optional
   }
 }
 
+void axis_check(IntArrayRef& axis, int64_t bound) {
+  std::unordered_set<int64_t> dimensions;
+  for (const auto i : c10::irange(axis.size())) {
+    int64_t pos_axis  = axis[i] >= 0 ? axis[i] : axis.size()+axis[i];
+    TORCH_CHECK( dimensions.find(pos_axis) == dimensions.end(), "torch.gradient requires axis to be unique");
+    TORCH_CHECK( (axis[i] < bound) && (axis[i] >= (-1*bound)), "torch.gradient's axis " + std::to_string(axis[i]) +" is out of bounds for array of dimension " + std::to_string(bound)  );
+    dimensions.insert( pos_axis);
+  }
+}
+
+std::vector<Tensor> gradient(const Tensor& self,
+                             TensorList spacing,
+                             IntArrayRef axis,
+                             int64_t edge_order) {
+  axis_check(axis, self.dim());
```

Comment:
This should also check that self and each tensor in spacing have the same device.

What about dtype? Should the dtypes of the tensors in self and spacing all be the same?

Reply:
Thanks for the comment, I have added check for device types. However, to my understanding, it's not essential spacing and and self having the same data type.